### Основные формулы комбинаторики

#### 0. Факториал

Произведение всех натуральных чисел от 1 до n.
$$
n! = 1 \times 2 \times 3 \times \dots \times n
$$
**Пример:** $5! = 1 \cdot 2 \cdot 3 \cdot 4 \cdot 5 = 120$

In [1]:
import math
print(f"5! = {math.factorial(5)}")

5! = 120


#### 1. Размещения (без повторений)
Упорядоченные выборки из n элементов по k.
$$
A_n^k = \frac{n!}{(n-k)!}
$$
**Пример:** Сколько способов выбрать капитана и заместителя из 10 человек?

$A_{10}^2 = \frac{10!}{8!} = 90$

In [8]:
print(f"A(10,2) = {math.perm(10, 2)}")

A(10,2) = 90


#### 2. Сочетания (без повторений)
Неупорядоченные выборки из n элементов по k (также известно как &laquo;це из эн по ка&raquo; и биномиальный коэффициент).
$$
C_n^k = \binom{n}{k} = \frac{n!}{k!(n-k)!}
$$
**Пример:** Сколькими способами выбрать 3 учеников из 10?


$C_{10}^3 = 120$

**Правило симметрии:**
$$
С_n^k = C_n^{n-k}
$$

In [3]:
print(f"C(10,3) = {math.comb(10, 3)}")

C(10,3) = 120


#### 3. Перестановки (без повторений)
Количество способов переставить n различных элементов.
$$
P_n = n!
$$
**Пример:** Сколькими способами можно расставить 5 книг?
$5! = 120$

In [5]:
print(f"P(5) = {math.factorial(5)}")

P(5) = 120


#### 4. Размещения с повторениями
Из n элементов выбираем k, порядок важен, элементы могут повторяться.
$$
\bar{A}_n^k = n^k
$$
**Пример:** Сколько существует 4-значных кодов из 10 цифр?

$10^4 = 10000$

In [26]:
print(f"A\u0304(10, 4): {10 ** 4}")

Ā(10, 4): 10000


#### 5. Сочетания с повторениями
Неупорядоченные выборки, элементы могут повторяться.
$$
\bar{C}_n^k = \binom{n+k-1}{k} =  \frac{(n+k-1)!}{k!(n-1)!}
$$
**Пример:** Сколькими способами можно выбрать 2 конфеты из 3 видов (конфеты одного вида одинаковы)?

$\bar{C}_3^2 = \binom{4}{2} = 6$

In [25]:
print(f"C\u0304(3, 2): {math.comb(3 + 2 - 1, 2)}")

C̄(3, 2): 6


### Задания для проверки

1. Сколько различных 6-значных чисел можно составить из цифр 1, 2, 3, 4, 5, 6, если цифры не повторяются?

2. В соревнованиях участвуют 12 спортсменов. Сколькими способами могут распределиться золотая, серебряная и бронзовая медали?

3. Из класса, в котором 15 человек, нужно выбрать 5 делегатов на конференцию. Сколькими способами это можно сделать?

4. Сколько существует 6-значных кодов из цифр 0-9? (код может начинаться с 0)

5. В магазине продаются пирожные 5 видов. Сколькими способами можно купить 8 пирожных?

6. Сколько различных трёхзначных чисел можно составить из цифр 1, 2, 3, 4, 5, если цифры в числе не повторяются? А сколько из них будут чётными?

7. В группе 20 студентов. Нужно выбрать старосту, его заместителя и 3 членов совета. Сколькими способами это можно сделать?

8. Из колоды в 36 карт (4 масти, 9 достоинств) нужно выбрать 5 карт. Сколькими способами это можно сделать, если среди выбранных карт:

А) Есть ровно 2 туза?

Б) Есть хотя бы 1 туз?

В) Есть ровно 3 карты одной масти?

9. Пароль должен состоять из 10 маленьких букв латинского алфавита и 6 цифр, при этом буквы не могут повторяться, а цифры могут. Буквы и цифры могут стоять в любом порядке. Сколько всего существует таких паролей?

### Модуль `itertools`

В модуле `itertools` есть несколько функций, полезных в работе с комбинаторикой.

#### Декартово произведение

Прямое произведение (декартово произведение) &mdash; множество, элементами которого являются все возможные упорядоченные последовательности элементов N исходных множеств длиной N. Эквивалентно N вложенным циклам. Вычисляется функцией `itertools.product()`.

**Пример:**

В столовой на первое можно выбрать борщ, рыбный суп или рассольник, на второе &mdash; котлету или сосиски, на гарнир &mdash; гречку, макароны или картошку. Сгенерируйте все варианты обеда.

In [7]:
import itertools

soups = ["борщ", "рыбный суп", "рассольник"]
second = ["котлета", "сосиски"]
side = ["гречка", "макароны", "картошка"]

for dinner in itertools.product(soups, second, side):
    print(dinner)

('борщ', 'котлета', 'гречка')
('борщ', 'котлета', 'макароны')
('борщ', 'котлета', 'картошка')
('борщ', 'сосиски', 'гречка')
('борщ', 'сосиски', 'макароны')
('борщ', 'сосиски', 'картошка')
('рыбный суп', 'котлета', 'гречка')
('рыбный суп', 'котлета', 'макароны')
('рыбный суп', 'котлета', 'картошка')
('рыбный суп', 'сосиски', 'гречка')
('рыбный суп', 'сосиски', 'макароны')
('рыбный суп', 'сосиски', 'картошка')
('рассольник', 'котлета', 'гречка')
('рассольник', 'котлета', 'макароны')
('рассольник', 'котлета', 'картошка')
('рассольник', 'сосиски', 'гречка')
('рассольник', 'сосиски', 'макароны')
('рассольник', 'сосиски', 'картошка')


**Пример 2:**

Из пяти первокурсников-гриффиндорцев каждый может привезти с собой сову, кошку или жабу или вообще не привозить домашнее животное. Какие возможны комбинации домашних животных?

Воспользуемся аргументом `repeat`.

In [8]:
pets = ["сова", "кошка", "жаба", "-"]

for p in itertools.product(pets, repeat=5):
    print(p)

('сова', 'сова', 'сова', 'сова', 'сова')
('сова', 'сова', 'сова', 'сова', 'кошка')
('сова', 'сова', 'сова', 'сова', 'жаба')
('сова', 'сова', 'сова', 'сова', '-')
('сова', 'сова', 'сова', 'кошка', 'сова')
('сова', 'сова', 'сова', 'кошка', 'кошка')
('сова', 'сова', 'сова', 'кошка', 'жаба')
('сова', 'сова', 'сова', 'кошка', '-')
('сова', 'сова', 'сова', 'жаба', 'сова')
('сова', 'сова', 'сова', 'жаба', 'кошка')
('сова', 'сова', 'сова', 'жаба', 'жаба')
('сова', 'сова', 'сова', 'жаба', '-')
('сова', 'сова', 'сова', '-', 'сова')
('сова', 'сова', 'сова', '-', 'кошка')
('сова', 'сова', 'сова', '-', 'жаба')
('сова', 'сова', 'сова', '-', '-')
('сова', 'сова', 'кошка', 'сова', 'сова')
('сова', 'сова', 'кошка', 'сова', 'кошка')
('сова', 'сова', 'кошка', 'сова', 'жаба')
('сова', 'сова', 'кошка', 'сова', '-')
('сова', 'сова', 'кошка', 'кошка', 'сова')
('сова', 'сова', 'кошка', 'кошка', 'кошка')
('сова', 'сова', 'кошка', 'кошка', 'жаба')
('сова', 'сова', 'кошка', 'кошка', '-')
('сова', 'сова', 'кошка'

#### Размещения

Вычисляются функцией `itertools.permutations()`.

**Пример:** найдите все возможные четырёхзначные коды из арабских цифр (0-9), в которых цифры не повторяются

In [10]:
for code in itertools.permutations("0123456789", 4):
    print("".join(code))

0123
0124
0125
0126
0127
0128
0129
0132
0134
0135
0136
0137
0138
0139
0142
0143
0145
0146
0147
0148
0149
0152
0153
0154
0156
0157
0158
0159
0162
0163
0164
0165
0167
0168
0169
0172
0173
0174
0175
0176
0178
0179
0182
0183
0184
0185
0186
0187
0189
0192
0193
0194
0195
0196
0197
0198
0213
0214
0215
0216
0217
0218
0219
0231
0234
0235
0236
0237
0238
0239
0241
0243
0245
0246
0247
0248
0249
0251
0253
0254
0256
0257
0258
0259
0261
0263
0264
0265
0267
0268
0269
0271
0273
0274
0275
0276
0278
0279
0281
0283
0284
0285
0286
0287
0289
0291
0293
0294
0295
0296
0297
0298
0312
0314
0315
0316
0317
0318
0319
0321
0324
0325
0326
0327
0328
0329
0341
0342
0345
0346
0347
0348
0349
0351
0352
0354
0356
0357
0358
0359
0361
0362
0364
0365
0367
0368
0369
0371
0372
0374
0375
0376
0378
0379
0381
0382
0384
0385
0386
0387
0389
0391
0392
0394
0395
0396
0397
0398
0412
0413
0415
0416
0417
0418
0419
0421
0423
0425
0426
0427
0428
0429
0431
0432
0435
0436
0437
0438
0439
0451
0452
0453
0456
0457
0458
0459
0461
0462
0463
0465


#### Сочетания

Вычисляются функцией `itertools.combinations()`.

**Пример:** найдите все способы выбрать двух дежурных из группы.

In [11]:
group = ["Петя", "Саша", "Вася", "Лиза", "Аня"]

for d1, d2 in itertools.combinations(group, 2):
    print(d1, d2)

Петя Саша
Петя Вася
Петя Лиза
Петя Аня
Саша Вася
Саша Лиза
Саша Аня
Вася Лиза
Вася Аня
Лиза Аня


#### Сочетания с повторениями

Вычисляются функцией `itertools.combinations_with_replacement()`.

**Пример:** в машине с напитками есть вода, фанта, спрайт, кока-кола и кока-кола лайт. Найдите все способы наполнить три стакана.

In [12]:
drinks = ["вода", "фанта", "спрайт", "кока-кола", "кока-кола лайт"]

for glasses in itertools.combinations_with_replacement(drinks, 3):
    print(glasses)

('вода', 'вода', 'вода')
('вода', 'вода', 'фанта')
('вода', 'вода', 'спрайт')
('вода', 'вода', 'кока-кола')
('вода', 'вода', 'кока-кола лайт')
('вода', 'фанта', 'фанта')
('вода', 'фанта', 'спрайт')
('вода', 'фанта', 'кока-кола')
('вода', 'фанта', 'кока-кола лайт')
('вода', 'спрайт', 'спрайт')
('вода', 'спрайт', 'кока-кола')
('вода', 'спрайт', 'кока-кола лайт')
('вода', 'кока-кола', 'кока-кола')
('вода', 'кока-кола', 'кока-кола лайт')
('вода', 'кока-кола лайт', 'кока-кола лайт')
('фанта', 'фанта', 'фанта')
('фанта', 'фанта', 'спрайт')
('фанта', 'фанта', 'кока-кола')
('фанта', 'фанта', 'кока-кола лайт')
('фанта', 'спрайт', 'спрайт')
('фанта', 'спрайт', 'кока-кола')
('фанта', 'спрайт', 'кока-кола лайт')
('фанта', 'кока-кола', 'кока-кола')
('фанта', 'кока-кола', 'кока-кола лайт')
('фанта', 'кока-кола лайт', 'кока-кола лайт')
('спрайт', 'спрайт', 'спрайт')
('спрайт', 'спрайт', 'кока-кола')
('спрайт', 'спрайт', 'кока-кола лайт')
('спрайт', 'кока-кола', 'кока-кола')
('спрайт', 'кока-кола', 'к

#### Разбор решения задачи с 15-ричными числами

Определите количество восьмизначных 15-⁠ричных чисел, в записи которых ровно два нуля и не более четырёх цифр, для записи которых используются буквы.

In [23]:
count = 0

# Ц - цифра (не 0), Б - буква, 0 - ноль

# сгенерируем "шаблоны" чисел
for i in itertools.product('0ЦБ', repeat=8):
    s = "".join(i)
    # если шаблон нас устраивает:
    # у него первая цифра не 0
    # и всего нулей 2
    # а букв <= 4
    if s[0] != "0" and s.count("0") == 2 and s.count("Б") <= 4:
        # то вычислим, сколько чисел соответствует шаблону
        # используя размещения с повторениями
        count += 9**s.count("Ц") * 5**s.count("Б")
print(count)

154248381


### Задания для самостоятельного выполнения

1. Найдите все пятизначные цифровые коды, в которых все цифры разные и сумма первых двух равна сумме последних двух. Запишите их в текстовый файл.

2. Найдите все возможные способы посадить в ряд двух мальчиков и трёх девочек так, чтобы мальчики не сидели рядом. Запишите их в текстовый файл.

3. В языке X 118 согласных и 267 гласных. Гласные в этом языке не могут стоять рядом, а сочетания согласных не могут превышать по длине 3. Сколько в этом языке возможно слов из 12 звуков, в которых все согласные разные (гласные могут повторяться)?


<details>
<summary>
Алгоритм решения:
</summary>

Назовём схемой слова последовательность из букв С и Г, которые отмечают, где в слове какие звуки. Например, для русского слова "ПРОСТО" схема будет "ССГССГ".

Для всех возможных схем длиной 12 определите, подходят ли они нам (если в них есть сочетания ГГ или СССС, то не подходят). Если подходят, то определите для каждой схемы, сколько возможно слов, подходящих под неё. Используйте функции из `math`. Чтобы определить, сколько в схеме гласных или согласных, используйте `count()`.
</details>

### Задания из ЕГЭ

1. Определите количество семизначных чисел, записанных в девятеричной системе счисления, в записи которых ровно одна цифра 6 и ровно две нечётные цифры.

2. Олег составляет таблицу кодовых слов для передачи сообщений, каждому сообщению соответствует своё кодовое слово. В качестве кодовых слов Олег использует 4-⁠буквенные слова, в которых есть только буквы A, B, C, D, E, X, Z, причём буквы X и Z встречаются только на двух первых позициях, а буквы A, B, C, D, E  — только на двух последних. Сколько различных кодовых слов может использовать Олег?

3. Сколько существует восьмеричных пятизначных чисел, не содержащих в своей записи цифру 1, в которых все цифры различны и никакие две чётные или две нечётные цифры не стоят рядом?

4. Ольга составляет 5-⁠буквенные коды из букв О, Л, Ь, Г, А. Каждую букву нужно использовать ровно 1 раз, при этом Ь нельзя ставить первым и нельзя ставить после гласной. Сколько различных кодов может составить Ольга?

5. Определите количество чисел, для записи которых в восьмеричной системе счисления требуется ровно 11 цифр, ровно 3 из которых  — нечётные, и никакие две нечётные цифры не стоят рядом.